File: 数据结构与算法/链表 linked list/0203. 移除链表元素/移除链表元素.py
"""
删除链表中等于给定值 val 的所有节点。

示例:

输入: 1->2->6->3->4->5->6, val = 6
输出: 1->2->3->4->5
"""
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeElements(self, head: ListNode, val: int) -> ListNode:
        if head is None:
            return None
        
        sentinel = ListNode(None)
        sentinel.next = head
        prev = sentinel
        cur = head

        while cur:
            if cur.val == val:
                prev.next = cur.next
            else:
                prev = cur
            cur = cur.next
        
        return sentinel.next

File: 数据结构与算法/链表 linked list/0203. 移除链表元素/移除链表元素.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        //添加虚拟头节点，迭代
        /*ListNode* res=new ListNode(NULL);
        res->next=head;
        ListNode* prev=res;
        ListNode* cur=head;
        while(cur!=NULL)
        {
            if(cur->val==val)   prev->next=cur->next;
            else    prev=cur;
            cur=cur->next;
        }
        return res->next;*/
        //递归
        if(head==NULL)
            return head;
        head->next=removeElements(head->next,val);
        if(head->val==val)
            return head->next;
        else
            return head;
    }
};